### PR TITLE
[Index] Walk implicit DeclRefExpr with an explicit ValueDecl.

### LINKED ISSUE #56189: [SR-13792] Index store should relate appendInterpolation from string literals
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-13792      |
   |Radar             | rdar://problem/70820224         |
   |Original Reporter | Leitch (JIRA User)      |
   |Type              | New Feature    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Source Tooling    |
   |Labels            | New Feature        |
   |Assignee          | Leitch (JIRA)      |
   |Priority          | Medium      |

   

   md5: 0896b2e76fc58c37c42a995dde600eea

  </details>

**Issue Description:**

Given the following code:

``` swift
extension String.StringInterpolation {
    mutating func appendInterpolation(test value: Int) {
        appendInterpolation(value)
    }
}

public struct Fixture112 {
    public func someFunc() {
        print("test: \(test: 1)")
    }
}
```

The index store could include a relation from the `someFunc()` occurrence to `appendInterpolation(test:)`. For the purposes of identifying unused code, it's not currently possible to identify that `appendInterpolation(test:)` is used in the string interpolation.

### PR DESCRIPTION
**Pull request by ileitch**
All references to explicit declarations should be included in the index store, even when those references are themselves implicit. For example, a custom `appendInterpolation(custom:)` method in `DefaultStringInterpolation` is referenced implicitly via string interpolation: `"\(custom: 1)"`.

Resolves: #56189

@benlangmuir

### CODE REVIEW

File: lib/IDE/SourceEntityWalker.cpp [hunk 1]
```diff
@@ -321,7 +328,8 @@ std::pair<bool, Expr *> SemaAnnotator::walkToExprPre(Expr *E) {
     }
   }
 
-  if (!isa<InOutExpr>(E) &&
+  if (!IsImplicitDeclRefToExplicitDecl &&
+      !isa<InOutExpr>(E) &&
```
Comment (ileitch) [inline lib/IDE/SourceEntityWalker.cpp:332]: Another option is to just add `!isa<DeclRefExpr>(E)` here, but I'm not sure of the implications of that, so opted for the safer option.

**Comment by benlangmuir** [comment]
@swift-ci please smoke test

File: lib/IDE/SourceEntityWalker.cpp (hunk 1 above)
Comment (benlangmuir) [reply lib/IDE/SourceEntityWalker.cpp:332]: I agree that checking for implicit references to explicit decls is the safe choice here.  There may be fully implicit cases that we want to handle, but my feeling is we're better off handling those on a case-by-case basis.

**Review by benlangmuir — APPROVED** [review]

**Comment by ileitch** [comment]
I didn't realize there were more index related tests outside of `test/Index`, so naturally there's a bunch of failures relating to SourceKit.

I wonder then, if this change should only apply to `IndexSwiftASTWalker`, to avoid any change to SourceKit behavior? Also one issue I just noticed is that because the `ValueDecl` is explicit, the created reference doesn't get the `Implicit` role. I could achieve that by adding a new visit method to `SourceEntityWalker` (`visitImplicitlyReferencedExplicitDecl`) similar to the other visit methods we have that handle special cases (e.g `visitCallAsFunctionReference`).

Also looks like I should be able to remove some/all of [`handleCustomAttrInitRefs`](https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L343-L365) once the `Implicit` role can be applied.

**Comment by bnbarham** [comment]
> I didn't realize there were more index related tests outside of `test/Index`, so naturally there's a bunch of failures relating to SourceKit.

The `SourceEntityWalker` is used for a bunch of different features, eg. cursor info.

> I wonder then, if this change should only apply to `IndexSwiftASTWalker`, to avoid any change to SourceKit behavior?

I think it's worth adding to `SemaAnnotator`. It would fix cursor info not having results for `test` in your added test (it should give the custom `appendInterpolation` method). Would need to add a few more tests for eg. cursor info and rename.

> Also one issue I just noticed is that because the `ValueDecl` is explicit, the created reference doesn't get the `Implicit` role. I could achieve that by adding a new visit method to `SourceEntityWalker` (`visitImplicitlyReferencedExplicitDecl`) similar to the other visit methods we have that handle special cases (e.g `visitCallAsFunctionReference`).

`passReference` takes a `ReferenceMetaData` where the third parameter is `isImplicit` (defaulting to false). You could change that to be `DRE->isImplicit()` where `passReference` is called in the `DRE` branch.

> Also looks like I should be able to remove some/all of [`handleCustomAttrInitRefs`](https://github.com/apple/swift/blob/main/lib/Index/Index.cpp#L343-L365) once the `Implicit` role can be applied.

I'm not sure I follow this part - what's the change that would make this possible?

**Comment by ileitch** [comment]
@benlangmuir @bnbarham @akyrtzi 

I'd like to get your thoughts on this.

```swift
class ClassA {
    init(){}
}
class ClassB {}
struct StructC {
    var x = 0
}
func someFunc() {
    _ = ClassA()
    _ = ClassB()
    _ = StructC()
    _ = StructC(x: 0)
}
```

Current `main` Swift will create references for `init()` for _all_ calls within `someFunc()`. However I think that behavior is wrong, as only `ClassA` has an explicit constructor declaration. The outcome for `ClassB` and `StructC` is misleading cursor info. Xcode's Jump To Definition lists `init()` as a destination, however since there's no explicit declaration the class/struct declaration is used as the location. If you look closely, you can see that the first 4 characters of the name are highlighted when performing the jump, which _would_ be correct if it were highlighting a constructor.

I'm proposing we remove these implicit references from the cursor info. Simply removing [CtorRefs](https://github.com/apple/swift/blob/22c6f9fb2c073426b1c1e0d78cf17de4c9f10afd/lib/IDE/SourceEntityWalker.cpp#L36) from `SemaAnnotator`, combined with the existing changes in this PR achieves that.

**Comment by ileitch** [comment]
> `passReference` takes a `ReferenceMetaData` where the third parameter is `isImplicit` (defaulting to false). You could change that to be `DRE->isImplicit()` where `passReference` is called in the `DRE` branch.

I gave that a try, and came to the conclusion that these references should remain explicit. If we make them implicit, they're excluded from cursor info. So while they're technically implicit, I think it make sense to treat them as explicit. From the developers perspective they _look_ explicit, since they're references to explicit declarations. Thoughts?

**Comment by benlangmuir** [comment]
Yeah, I think the *reference* should be explicit here, even if the underlying decl is not.  Just like a reference to a synthesized function (e.g. from Hashable or Codable synthesis) ought to be an explicit reference.

**Comment by bnbarham** [comment]
> Yeah, I think the _reference_ should be explicit here, even if the underlying decl is not. Just like a reference to a synthesized function (e.g. from Hashable or Codable synthesis) ought to be an explicit reference.

So I definitely agree that any actual reference (whether to an explicit or implicit decl) should be explicit. I'm not so sure about the case in the PR though - it's fairly borderline as to whether it's an implicit reference or not. The label is explicit but the call really does seem implicit to me.

If the only issue is that cursor info ignores implicit references but we want it do provide results for this case, I'd argue that we should just allow cursor info to visit them. I could be wrong, but my guess is that they were initially ignored because they sometimes don't have a source location, which would break `CursorInfoResolver::visitDeclReference` since it relies on having a valid range. I recently changed that to actually check for a valid range anyway.

Happy to take counter examples, but for the couple I can think of it would actually be useful to provide implicit references in the results - that applies for the `appendInterpolation` case as well as dynamic member lookups (eg. `test/SourceKit/CursorInfo/cursor_info_keypath_member_lookup.swift:25:9` could give `subscript` as a result as well as `topLeft`). Note that I haven't actually checked whether we give a location in that case, we could always add that though 🤷

**Comment by benlangmuir** [comment]
>  I'd argue that we should just allow cursor info to visit them. I could be wrong, but my guess is that they were initially ignored because they sometimes don't have a source location, which would break CursorInfoResolver::visitDeclReference since it relies on having a valid range. I recently changed that to actually check for a valid range anyway.

SGTM

**Comment by bnbarham** [comment]
> Perhaps a way forward is to remove the explicit/implicit flag from references, and instead have cursor info look at the referenced decl when deciding if it should be included or not. I could be wrong, but I think in that case we'd only ever want to include explicit decls? What about others feature, e.g refactoring?

I don't think we want to remove that flag from the index. I'd go ahead with what I said above, ie. just remove the implicit check on the reference in the cursor info resolver.

**Comment by ileitch** [comment]
> I don't think we want to remove that flag from the index. I'd go ahead with what I said above, ie. just remove the implicit check on the reference in the cursor info resolver.

Might that then include things like implicit property accessor decls in cursor info? I believe they have the same source location as the property themselves, so it sounds like they might slip past your change to `CursorInfoResolver::visitDeclReference`?

Using the explicitness of the referenced decl would avoid cases like this I think. Is there a scenario in which we'd actually want a reference to an _implicit_ decl included in cursor info?

**Comment by ileitch** [comment]
Also I'm still confused as to why we'd want explicit references to implicit constructors: https://github.com/apple/swift/pull/38900#issuecomment-986236161. This appears to have been an intentional change, added in https://github.com/apple/swift/commit/40c8904.

**Comment by bnbarham** [comment]
> Might that then include things like implicit property accessor decls in cursor info? I believe they have the same source location as the property themselves, so it sounds like they might slip past your change to `CursorInfoResolver::visitDeclReference`?
>
> Using the explicitness of the referenced decl would avoid cases like this I think. Is there a scenario in which we'd actually want a reference to an _implicit_ decl included in cursor info?

I think it's reasonable for cursor info to return all involved symbols, whether implicit or not. Clients can then decide themselves whether to show them, ie. they could ignore symbols flagged with "synthesized" if there are non-synthesized symbols as well. An example there could be `==` and `hashValue` where IMO it would make sense to go to the location that caused them to be synthesized (ie. wherever `Equatable`/`Hashable` were added), though right now we don't currently add locations to them.

**Comment by akyrtzi** [comment]
> Also I'm still confused as to why we'd want explicit references to implicit constructors

The discussion here focuses on what cursor-info should do, but AFAIK the way we walk the AST affects what symbol occurrences we get for indexing purposes. For the indexing data we need all the symbol references whether the declarations are implicit or not.

For example, for the case of implicit constructors we want the capability to do an index query to find all the places where that particular type is constructed.

**Comment by ileitch** [comment]
> For example, for the case of implicit constructors we want the capability to do an index query to find all the places where that particular type is constructed.

Thanks, that makes sense. I'd rephrase my question a little then. Why do the references to the implicit constructors not have their implicit flag set? It has the unfortunate side effect of those references being included in cursor info, which causes Xcode to offer up the constructor as a jump to destination.

If it's OK for those references to marked as implicit, it'd simplify the changes I need to make in this PR.

**Comment by bnbarham** [comment]
> Thanks, that makes sense. I'd rephrase my question a little then. Why do the references to the implicit constructors not have their implicit flag set? It has the unfortunate side effect of those references being included in cursor info, which causes Xcode to offer up the constructor as a jump to destination.
> 
> If it's OK for those references to marked as implicit, it'd simplify the changes I need to make in this PR.

The reference is explicit, the symbol is implicit/synthesized - which is now marked with `synthesized` in cursor info. The references *should* be included in cursor info. It's up to clients of cursor info as to what they do with that information.

**Comment by ileitch** [comment]
Sorry you're right, I should have said that the reference being explicit was causing Xcode to display it, not merely it's presence in cursor info.

Is the synthesized in cursor info brand new? I didn't see that when I pulled main on the weekend.

**Comment by bnbarham** [comment]
> Sorry you're right, I should have said that the reference being explicit was causing Xcode to display it, not merely it's presence in cursor info.

New-ish but not brand new: https://github.com/apple/swift/pull/40062
